brian2/codegen/runtime/cython_rt/templates/spikegenerator.pyx

```diff
@@ -1,39 +1,45 @@
 {% extends 'common.pyx' %}
 
-{% block template_support_code %}
-cdef int _start_idx = 0
-cdef double _current_time = 0.0
-{% endblock %}
-
 {% block maincode %}
 
-    {# USES_VARIABLES {_spikespace, N, t, dt, neuron_index, spike_time } #}
-     
-    global _start_idx, _current_time
-    
-    # Check if we have been reset, if so reset the cache
-    if t<_current_time:
-        _start_idx = 0
-    
-    _current_time = t
+    {# USES_VARIABLES {_spikespace, N, t, dt, neuron_index, spike_time, period, _lastindex } #}
 
     # TODO: We don't deal with more than one spike per neuron yet

```

Comment:
About doing the conversion to integer time steps at the preparation stage: I'm wondering if we could do this in an elegant way, also incorporating the offset calculation we had in Brian 1.
The simple integer conversion would convert (for dt=0.1ms)
```
indices = [0, 1, 0, 1, 0, 1]
times = [0.1, 0.1, 0.2, 0.3, 1.0, 1.0] * ms
```
into
```
timebins = [1, 1, 2, 3, 10, 10]
```
In Brian 1, we'd calculate offsets:
```
offsets = [0, 0, 2, 3, 4, 4, 4, 4, 4, 4, 4, 6]
```
The Brian 1 solution isn't great for long runs with few spikes, because it stores an array of the size of the number of time steps. But we could a similar thing by just storing the number of spikes falling in the same time bin, but only for spikes that actually happen:
```
same_bin_spikes = [2, 1, 1, 2]
```
I think this could be calculated quite efficiently before hand and would simplify the templates significantly, in numpy we'd have to do neither looping nor bisection during runtime.